### PR TITLE
react rule state-in-constructor

### PR DESCRIPTION
**Pull request by aravind1078**
Rule react/state-in-constructor

#341

### CODE REVIEW

**Comment by ematipico** [comment]
@aravind1078 thank you for your PR! You might want to run `scripts/dev-rome check --apply` and commit the changes!

**Comment by aravind1078** [comment]
> @aravind1078 thank you for your PR! You might want to run `scripts/dev-rome check --apply` and commit the changes! This should fix some problems occurring in the CI

yup just noticed it...

**Comment by aravind1078** [comment]
@ematipico looks like there is one more problem with generated files check... can I do something to fix that ?

**Comment by ematipico** [comment]
It looks good now! Great job!

File: packages/@romejs/compiler/lint/rules/react/stateInConstructor.test.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+				            state = { bar: 0 }
+				            render() {
+				                return <div>Foo</div>
+				            }
+				        }
+					`,
+				],
+				valid: [
+					dedent`
+						class Foo extends React.Component {
```
Comment (ematipico) [inline packages/@romejs/compiler/lint/rules/react/stateInConstructor.test.ts:23]: Could we have another case where we read the `state` somewhere else? To make sure we capture the correct instance of `state`

```js
class Foo extends Component {
   	constructor() {
		this.state { text: "foo" }
	}

	render() {
		const { state } = this;
		const { text } = state;
		return (
			<span>{text}</span>
		)
	}	
}

```

Comment (aravind1078) [reply packages/@romejs/compiler/lint/rules/react/stateInConstructor.test.ts:23]: yeah sure. I will add that. Thanks!

**Comment by sebmck** [comment]
Hey! Thanks for the PR! I'm not sure if it makes sense for us to have this rule. We haven't yet figured out what patterns we want to encourage and I think this one is too restrictive. Other core contributors may have some thoughts. Happy to defer.

**Comment by aravind1078** [comment]
> Hey! Thanks for the PR! I'm not sure if it makes sense for us to have this rule. We haven't yet figured out what patterns we want to encourage and I think this one is too restrictive. Other core contributors may have some thoughts. Happy to defer.

Yeah it does seem but i think its good to have uniformity across the project it could be inside/outside the constructor. React-eslint does provide that flexibility to toggle between inside and outside the constructor but am not sure whether we can do that here. Would like to hear what others think. Thoughts ? @ematipico @bitpshr @VictorHom

**Comment by sebmck** [comment]
Rome doesn’t have any lint options and I don’t think we intend on having them, which is why making these decisions is so difficult since well be forcing Rome users into these patterns.

**Comment by ematipico** [comment]
Personally, I would like to have the choice. Forcing the definition of `this.state` inside the `constructor` would force to always implement `super()` even when it is not necessary. 

Usually I define `state` inside the `constructor` only if I need to compute the `state` from the `props`. Otherwise, as a property of the class is fine. Does this lead to errors or is it just code style? If it is just the latter, I would not enforce this rule.

**Comment by bitpshr** [comment]
After thinking about this a bit more, I also feel that this may be too restrictive of a rule. Apologies for not calling this out sooner @aravind1078.

It's difficult to justify subjectively initializing state in the `constructor` when initial values are not computed and don't actually require initialization. In these cases, it feels like using class properties would be the preferred style choice. For example, the following feels less error-prone than overriding the `constructor` and remembering to properly call `super`:

```ts
class Foo extends React.Component {
    state = { open: true };
   // ...
}
```

Like @sebmck said, Rome has no configuration and can be thought of as an "all or nothing" code standardizer, so subjective rules should be limited to only those with wide community support. Unless there are objective reasons to encourage `constructor`-based initialization beyond consistency with other codebases, I feel that we should close this rule for now and revisit it in the future if there's ever community desire.

**Comment by aravind1078** [comment]
Sure no problem. Thanks for your inputs. I will close this pr for now then. Since react-rules are now paused. Please let me know if there is something open for contributions. Eager and Happy to contribute. Cheers!!

**Pull request closed without merging**